cluster_pack/uploader.py

```diff
@@ -174,32 +185,49 @@ def _get_hash(spec_file: str) -> str:
 
 def _upload_zip(
     zip_file: str, package_path: str,
-    resolved_fs: Any = None, force_upload: bool = False
+    resolved_fs: filesystem.EnhancedFileSystem, force_upload: bool = False
 ) -> None:
     packer = packaging.detect_packer_from_file(zip_file)
     if packer == packaging.PEX_PACKER and resolved_fs.exists(package_path):
-        with tempfile.TemporaryDirectory() as tempdir:
-            local_copy_path = os.path.join(tempdir, os.path.basename(package_path))
-            resolved_fs.get(package_path, local_copy_path)
-            info_from_storage = PexInfo.from_pex(local_copy_path)
-            into_to_upload = PexInfo.from_pex(zip_file)
-            if not force_upload and info_from_storage.code_hash == into_to_upload.code_hash:
-                _logger.info(f"skip upload of current {zip_file}"
-                             f" as it is already uploaded on {package_path}")
-                return
+        if not force_upload and _check_pex_version(resolved_fs, zip_file, package_path):
+            _logger.info(f"skip upload of current {zip_file}"
+                         f" as it is already uploaded on {package_path}")
+            return
+        elif resolved_fs.exists(package_path):

```

Comment:
Was it this alreayd checked L191? If we are here I guess we know that the pex was already uploaded at this path but the pex versions are mismatching.

Plus I guess here you are removing the pex L198 because otherwise the put_atomic fails (just to be sure I got it) ?

Reply:
Yes indeed. I will simplify this: after a review with Marc, the expected behavior is:
- fail if pex exists with the wrong version (unless force upload)
- remove existing only if force upload
BTW this will be the only code path modified, as other path were not subject to the race condition (see internal discussion on slack)